IdentityCore/src/requests/MSIDSilentTokenRequest.m

```diff
@@ -235,129 +216,145 @@ - (void)executeRequestImpl:(MSIDRequestCompletionBlock)completionBlock
-                 }
-                 else
-                 {
-                     completionBlock(result, nil);
-                 }
-             }];
+    BOOL isAppRefreshToken = tokenType == MSIDAppRefreshTokenType;
+    MSID_LOG_WITH_CTX_PII(MSIDLogLevelVerbose, self.requestParameters, @"Trying to acquire access token using %@ for clientId %@, authority %@, account %@", (!isAppRefreshToken ? @"App Refresh Token" : @"Family Refresh Token"), self.requestParameters.authority, self.requestParameters.clientId, self.requestParameters.accountIdentifier.maskedHomeAccountId);
+    
+    // When using ART or FRT, it will go through the same method below, and handle differently within the completion block
+    [self redeemAccessTokenWith:refreshToken
+                completionBlock:^(MSIDTokenResult * _Nullable result, NSError * _Nullable error) {
+        if (error)
+        {
+            if ([self isErrorRecoverableByUserInteraction:error])
+            {
+                if (isAppRefreshToken)
+                {
+                    [self handleErrorResponseForAppRefreshToken:refreshToken completionBlock:completionBlock];
+                }
+                else
+                {
+                    [self handleErrorResponseForFamilyRefreshToken:error completionBlock:completionBlock];
+                }
+                
+                NSError *interactionError = MSIDCreateError(MSIDErrorDomain, MSIDErrorInteractionRequired, @"User interaction is required", error.msidOauthError, error.msidSubError, error, self.requestParameters.correlationId, nil, YES);
+                completionBlock(nil, interactionError);
+            }
+            else
+            {
+                completionBlock(nil, error);
+            }
+        }
+        else
+        {
+            completionBlock(result, nil);
+        }
+    }];
 }
 
-- (void)tryAppRefreshToken:(MSIDBaseToken<MSIDRefreshableToken> *)multiResourceRefreshToken
-           completionBlock:(nonnull MSIDRequestCompletionBlock)completionBlock
+#pragma mark - Helpers
+
+- (void)handleErrorResponseForAppRefreshToken:(MSIDBaseToken<MSIDRefreshableToken> *)refreshToken
+                              completionBlock:(nonnull MSIDRequestCompletionBlock)completionBlock
 {
-    if (!multiResourceRefreshToken)
+    NSError *error = nil;
+    MSIDRefreshToken *familyRefreshToken = [self familyRefreshTokenWithError:&error];
+    if (error)
     {
-        MSID_LOG_WITH_CTX(MSIDLogLevelInfo, self.requestParameters, @"Refresh token wasn't found, user interaction is required.");
-        
-        NSError *interactionError = MSIDCreateError(MSIDErrorDomain, MSIDErrorInteractionRequired, @"User interaction is required", nil, nil, nil, self.requestParameters.correlationId, nil, YES);
-        completionBlock(nil, interactionError);
+        MSID_LOG_WITH_CTX_PII(MSIDLogLevelError, self.requestParameters, @"Failed to retrieve Family Refresh token with error %@", MSID_PII_LOG_MASKABLE(error));
```

Comment:
Should we return interaction required here instead? User would then at least be able to recover by acquiring tokens interactively. With this error user would be blocked.

Reply:
Reworked to take the same interactive flow if there is no FRT available